x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,419 @@
+                int roundedAllocations = random.nextDouble() < roundUpProbability
+                    ? (int) Math.ceil(softAllocations.get(assignment))
+                    : (int) Math.floor(softAllocations.get(assignment));
+
+                if (m.memoryBytes() > resourceTracker.remainingNodeMemory.get(n)
+                    || m.threadsPerAllocation() > resourceTracker.remainingNodeCores.get(n)
+                    || roundedAllocations == 0
+                    || random.nextDouble() > softAssignments.get(assignment)) {
+                    softAssignments.put(assignment, 0.0);
+                    softAllocations.put(assignment, 0.0);
+                    assignUnderSubscribedNodes(Set.of(n));
+                } else {
+                    roundedAllocations = Math.min(roundedAllocations, resourceTracker.remainingNodeCores.get(n) / m.threadsPerAllocation());
+                    assignModelToNode(m, n, roundedAllocations);
+                    unassignOversizedModels(n);
+                    assignExcessCores(n);
+                }
+            }
+        }
+
+        private void unassignOversizedModels(Node n) {
+            for (Model m : models) {
+                Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+                if (softAssignments.get(assignment) < 1.0 && m.memoryBytes() > resourceTracker.remainingNodeMemory.get(n)) {
+                    softAssignments.put(assignment, 0.0);
+                    softAllocations.put(assignment, 0.0);
+                }
+            }
+        }
+
+        private AssignmentPlan toPlan() {
+            AssignmentPlan.Builder builder = AssignmentPlan.builder(nodes, models);
+            for (Map.Entry<Tuple<Model, Node>, Integer> assignment : tryAssigningRemainingCores().entrySet()) {
+                builder.assignModelToNode(assignment.getKey().v1(), assignment.getKey().v2(), assignment.getValue());
+            }
+            return builder.build();
+        }
+
+        private Map<Tuple<Model, Node>, Integer> tryAssigningRemainingCores() {
+            // Eagerly assign allocations to models with larger size first on first node
+            // where the model fits.
+            //
+            // This is a trivial way to improve solution quality since increasing
+            // used allocations always improves our quality measure and we may be able to
+            // add a job, which doesn't have its quota of allocations, to the allocation
+            // random rounding finds.
+
+            Map<Tuple<Model, Node>, Integer> resultAllocations = new HashMap<>();
+
+            ResourceTracker resourceTracker = new ResourceTracker(nodes, models);
+
+            for (Model m : models) {
+                for (Node n : nodes) {
+                    Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+                    int allocations = (int) Math.floor(softAllocations.getOrDefault(assignment, 0.0));
```

Comment:
OK, I think this `softAllocations` and `softAllocations` all needs to be wrapped up. Instead of `put(0.0)` can we remove the key? 

And not access those maps directly except through specific methods? It seems like "knowledge of how to access values and update them, etc." is bleeding throughout with these maps and it may be easier to follow if they were contained somehow.

Reply:
Actually the `getOrDefault` is unnecessary here. There should always be a value for all possible `(m, n)`.

`AssignmentHolder` is all about interacting with those values. Not sure how I could wrap them further. I'll have a think.

Reply:
@dimitris-athanasiou it just seems that every time the interaction takes place, the same logic is used: make tuple, get it, check if its zero or something. Additionally, it seems that almost in every case, both maps need to be updated.

I am thinking that this should be wrapped up in a method or something. Additionally, storing zero in the maps seems weird. Why not delete keys and treat `0` as the default value?

Reply:
I have extracted a method for unassigning a soft assignment in 437ec1b80fd1122a100734b40a39ab2845e6d04b.

I'd rather we kept `0` values in there as otherwise we'd have to be doing `contains` checks in a bunch of places.

Reply:
> I'd rather we kept 0 values in there as otherwise we'd have to be doing contains checks in a bunch of places.

You shouldn't have to. `getOrDefault` is exactly what you want.